lib/core/src/sync/mod.rs

```diff
@@ -435,27 +440,38 @@ impl SyncService {
         let record = Record::new(sync_data, record_revision, self.signer.clone())?;
         trace!("Got record: {record:?}");
 
-        // Step 4: Push the record
-        let req = SetRecordRequest::new(record, utils::now(), self.signer.clone())?;
-        trace!("Got set record request: {req:?}");
-        let reply = self.client.push(req).await?;
-        trace!("Got reply: {reply:?}");
+        // Step 4: Lock record, so we can track pending outbound changes
+        self.persister
+            .lock_outgoing_record(data_id, record_id, is_local)?;
+
+        if let Err(err) = {
+            // Step 5: Push the record
+            let req = SetRecordRequest::new(record, utils::now(), self.signer.clone())?;
+            trace!("Got set record request: {req:?}");
+            let reply = self.client.push(req).await?;
+            trace!("Got reply: {reply:?}");
+
+            // Step 6: Check for conflict. If present, skip and retry on the next call
+            if reply.status() == SetRecordStatus::Conflict {
+                return Err(anyhow!(
+                    "Got conflict status when attempting to push record"
+                ));
+            }
 
-        // Step 5: Check for conflict. If present, skip and retry on the next call
-        if reply.status() == SetRecordStatus::Conflict {
-            return Err(anyhow!(
-                "Got conflict status when attempting to push record"
-            ));
+            // Step 7: Set/update the state revision
+            self.persister.set_sync_state(SyncState {
+                data_id: data_id.to_string(),
+                record_id: record_id.to_string(),
+                record_revision: reply.new_revision,
+                is_local,
+            })?;
+
+            Ok(())
+        } {
+            self.persister.unlock_outgoing_record(record_id)?;
```

Comment:
What happen if we exit before executing this line? Seems like the record is locked forever?